### ISSUE TITLE
Add support for "option" field in protobuf message

### ISSUE THREAD

**Issue by idokleinman** [apple/swift-protobuf#761]
For a protobuf message of type

```
message AuthRequest {
  option (type_id) = 1001;
  // Commissioning credential
  string password = 1;
}
```

Resulting Swift code does not contain the `option (type_id) = 1001;` and the number cannot be accessed. Are there plans to add support for class descriptions (aka protobuf option)?
Is it already supported and I'm missing a plugin switch?

**Comment by seivan** [comment]
@idokleinman Protobuf 3 made everything optional by default and got rid of the notion of 'required' (the opposite of optional).

Now, there are arguments against and for it, but I don't think the generated code can detect if something is optional if it's been set with a default value. In fact, it's an issue I'm dealing with here, trying to figure out if a value is default set or set since it's optional, it's hard to tell. 

It's a step down for some, and a step up for others. I personally would have preferred it, but not sure if `swift-protobuf` supports Protobuf 2 specs and I rather not go back a version that *might* be abandoned (Google claims they won't).

**Comment by seivan** [comment]
I misunderstood, and thought you meant optional.

**Comment by thomasvl** [comment]
Options are metadata that is only captured in the _descriptor_ describing the messages.  The Swift proto runtime currently doesn't need any of this information because doesn't use it and the data can be a non trivial amount of data added to binaries that most things don't need.

The first step would likely be to bundle the generated code for _descriptor.proto` with the library (Issue #727), and then see about how it makes sense to expose this from messages.

**Comment by richardasmall** [comment]
Has there been any more thought put into this? Just discovered this myself. Our team uses custom options extensively in our protos (for better or worse).

**Comment by thomasvl** [comment]
Not really.  We did start bundling `descriptor.proto` a while ago to remove that issue.

As mentioned, since the data can be non trivial in size, figuring out a way to do this that doesn't bloat things is important.  The other issues asking for the protobuf _reflection_ type apis (act on message fields generically), also need some of the same information/apis, so there's a good chance that would all be done together.

**Comment by tbkka** [comment]
Other protobuf implementations handle this by dumping the serialized descriptors into the generated code in their entirety.  When you access your custom option, they deserialize the descriptor and look up the option data there.

If this is important to you, you can do this manually:  Use `protoc` to dump the descriptors, bundle them into your code, and then deserialize them and inspect them to retrieve the custom options.  It's a little tedious, but feasible.

We've been slow to implement this by default for several reasons:
1. Swift does not currently have a really good way to embed binary data in the generated code.
2. The deserialized descriptors can use a lot of memory.  Many people object to this.
3. Getting useful data out of the descriptors can be pretty tedious.  Other implementations have pretty involved collections of helper code to present a nicer interface to this.  We've built a little of that support for the code generator but it's not yet complete enough to support these kinds of uses.

**Comment by tbkka** [comment]
Here are some hints for folks who want to use options:

1. You can generate a `FileDescriptorSet` using protoc:
```
  protoc --descriptor_set_out=.  some.proto other.proto
```

2. That `FileDescriptorSet` is a protobuf-serialized copy of your proto definitions.  You can deserialize it with the descriptor type that's included in the SwiftProtobuf library:
```
 import SwiftProtobuf
 let fileDescSet = Google_Protobuf_FileDescriptorSet(serializedData: data)
```

3. From this, you can walk the proto definition and read off any options you're interested in.  The details of the file descriptor set are in `descriptor.proto` here:  https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto

In particular:
* A "FileDescriptorSet" holds an array of "FileDescriptor" objects, one for each file in the set.
* A "FileDescriptor" object has an array of "Descriptor" objects defining each message.
* A "Descriptor" has an array of "FieldDescriptor" objects defining the fields of that message.
* Each "Descriptor" and "FieldDescriptor" includes an "options" array with all options defined on that message or field.

The above information should be enough to make use of option data.  The only missing piece is somehow embedding the information into your project.  The C++ implementation does this by embedding the entire FileDescriptorSet into the generated output, and implicitly deserializing the entire thing in order to look up option information.  This has obvious problems with size (code size for the embedded data and memory size for the decoded descriptor information) which is a major reason we've been hesitant to implement this so far.

If enough people experiment with this, maybe we can collectively acquire enough experience with different approaches to come up with a good solution.  (For example, I've recently started wondering if there could be a separate generator that just generated this information as extensions to the Swift message types; you could then run that separate generator in addition to `protoc-gen-swift` to get this additional support.  A similar approach might be used to split out the JSON/text support logic that some people would like to omit.)
